group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -544,6 +550,77 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    /**
+     * Remove expired offsets for group.
+     *
+     * @param groupId The group id.
+     * @param records The list of records to populate with offset commit tombstone records.
+     *
+     * @return True if no offsets exist or if all offsets expired, false otherwise.
+     */
+    public boolean cleanupExpiredOffsets(String groupId, List<Record> records) {
+        TimelineHashMap<String, TimelineHashMap<Integer, OffsetAndMetadata>> offsetsByTopic = offsetsByGroup.get(groupId);
+        if (offsetsByTopic == null) {
+            return true;
+        }
+
+        // We expect the group to exist.
```

Comment:
Is this method only called after the groupId is verified to exist?

Reply:
yes

```
        groupMetadataManager.groupIds().forEach(groupId -> {
            if (offsetMetadataManager.cleanupExpiredOffsets(groupId, records)) {
                groupMetadataManager.maybeDeleteGroup(groupId, records);
            }
        });
```